cirq-core/cirq/ops/gateset.py

```diff
@@ -0,0 +1,378 @@
+        """Returns a copy of this Gateset with identical gates and new values for named arguments.
+
+        If a named argument is None then corresponding value of this Gateset is used instead.
+
+        Args:
+            name: New name for the Gateset.
+            unroll_circuit_op: If True, new Gateset will recursively validate
+                `cirq.CircuitOperation` by validating the underlying `cirq.Circuit`.
+            accept_global_phase: If True, new Gateset will accept `cirq.GlobalPhaseOperation`.
+
+        Returns:
+            `self` if all new values are None or identical to the values of current Gateset.
+            else a new Gateset with identical gates and new values for named arguments.
+        """
+
+        def val_if_none(var: Any, val: Any) -> Any:
+            return var if var is not None else val
+
+        name = val_if_none(name, self._name)
+        unroll_circuit_op = val_if_none(unroll_circuit_op, self._unroll_circuit_op)
+        accept_global_phase = val_if_none(accept_global_phase, self._accept_global_phase)
+        if (
+            name == self._name
+            and unroll_circuit_op == self._unroll_circuit_op
+            and accept_global_phase == self._accept_global_phase
+        ):
+            return self
+        return Gateset(
+            *self.gates,
+            name=name,
+            unroll_circuit_op=cast(bool, unroll_circuit_op),
+            accept_global_phase=cast(bool, accept_global_phase),
+        )
+
+    def __contains__(self, item: Union[raw_types.Gate, raw_types.Operation]) -> bool:
+        """Check for containment of a given Gate/Operation in this Gateset.
+
+        Containment checks are handled as follows:
+            a) For Gates or Operations that have an underlying gate (i.e. op.gate is not None):
+                - Forwards the containment check to the underlying GateFamily's
+                - Examples of such operations include `GateOperations` and their controlled and
+                    tagged variants (i.e. instances of `TaggedOperation`, `ControlledOperation`
+                    where `op.gate` is not None) etc.
+
+            b) For Operations that do not have an underlying gate:
+                - Forwards the containment check to `self._validate_operation(item)`.
+                - Examples of such operations include `CircuitOperations` and their controlled
+                    and tagged variants (i.e. instances of `TaggedOperation`, `ControlledOperation`
+                    where `op.gate` is None) etc.
+
+        The complexity of the method is:
+            a) O(1) for checking containment in the default `cirq.GateFamily` instances.
+            b) O(n) for checking containment in custom GateFamily instances.
+
+        Args:
+            item: The `cirq.Gate` or `cirq.Operation` instance to check containment for.
+        """
+        if isinstance(item, raw_types.Operation) and item.gate is None:
+            return self._validate_operation(item)
```

Comment:
Should this call into `_validate_operation` or `validate` ?

Reply:
I've assumed here that the user will respect the function signature and not pass in an `op_tree` or `circuit` to check for containment in the Gateset. Hence, calling `_validate_operation` is the right thing to do. 

The recommendation in `_validate_operation` is also that:
"Users should override this method to define custom behavior for operations that do not have an underlying `cirq.Gate`."